qutip/nonmarkov/bofin_solvers.py

```diff
@@ -0,0 +1,1218 @@
+        if self.ados.exponents[k].type in (
+                BathExponent.types.R, BathExponent.types.I,
+                BathExponent.types.RI
+        ):
+            return self._grad_prev_bosonic(he_n, k)
+        elif self.ados.exponents[k].type in (
+                BathExponent.types["+"], BathExponent.types["-"]
+        ):
+            return self._grad_prev_fermionic(he_n, k)
+        else:
+            raise ValueError(
+                f"Mode {k} has unsupported type {self.ados.exponents[k].type}")
+
+    def _grad_prev_bosonic(self, he_n, k):
+        if self.ados.exponents[k].type == BathExponent.types.R:
+            op = (-1j * he_n[k] * self.ados.ck[k]) * (
+                self._s_pre_minus_post_Q[k]
+            )
+        elif self.ados.exponents[k].type == BathExponent.types.I:
+            op = (-1j * he_n[k] * 1j * self.ados.ck[k]) * (
+                self._s_pre_plus_post_Q[k]
+            )
+        elif self.ados.exponents[k].type == BathExponent.types.RI:
+            term1 = (he_n[k] * -1j * self.ados.ck[k]) * (
+                self._s_pre_minus_post_Q[k]
+            )
+            term2 = (he_n[k] * self.ados.ck2[k]) * self._s_pre_plus_post_Q[k]
+            op = term1 + term2
+        else:
+            raise ValueError(
+                f"Unsupported type {self.ados.exponents[k].type}"
+                f" for exponent {k}"
+            )
+        return op
+
+    def _grad_prev_fermionic(self, he_n, k):
+        ck = self.ados.ck
+
+        n_excite = sum(he_n)
+        sign1 = (-1) ** (n_excite + 1)
+
+        n_excite_before_m = sum(he_n[:k])
+        sign2 = (-1) ** (n_excite_before_m)
+
+        sigma_bar_k = k + self.ados.sigma_bar_k_offset[k]
+
+        op = -1j * sign2 * (
+            (ck[k] * self._spreQ[k]) -
+            (sign1 * np.conj(ck[sigma_bar_k] * self._spostQ[k]))
+        )
+
+        return op
+
+    def _grad_next(self, he_n, k):
+        """ Get the previous gradient. """
+        if self.ados.exponents[k].type in (
+                BathExponent.types.R, BathExponent.types.I,
+                BathExponent.types.RI
+        ):
+            return self._grad_next_bosonic(he_n, k)
+        elif self.ados.exponents[k].type in (
+                BathExponent.types["+"], BathExponent.types["-"]
+        ):
+            return self._grad_next_fermionic(he_n, k)
+        else:
+            raise ValueError(
+                f"Mode {k} has unsupported type {self.ados.exponents[k].type}")
+
+    def _grad_next_bosonic(self, he_n, k):
+        op = -1j * self._s_pre_minus_post_Q[k]
+        return op
+
+    def _grad_next_fermionic(self, he_n, k):
+        n_excite = sum(he_n)
+        sign1 = (-1) ** (n_excite + 1)
+
+        n_excite_before_m = sum(he_n[:k])
+        sign2 = (-1) ** (n_excite_before_m)
+
+        if sign1 == -1:
+            op = (-1j * sign2) * self._s_pre_minus_post_Q[k]
+        else:
+            op = (-1j * sign2) * self._s_pre_plus_post_Q[k]
+
```

Comment:
I think in 601 and 604 the Q operators should be the adjoint (dagger) of the Q associated with that exponent

Reply:
I agree that these should be based on `Q.dag()` or equivalently taken from `sigma_bar_k` instead of `k`. Do you know which of those two ways of expressing this is best?

After I fixed this, the currents in the test ended up being 0, which doesn't match the analytic result. Any thoughts on what to try next?

Reply:
I think there is also a separate question about what `exp.Q` should be for the `+` and `-` exponent types.

Currently, if the operator `Q` is supplied to the bath, then the + modes have `exp.Q == Q.dag()` and the - modes have `exp.Q == Q`. Perhaps the `exp.Q` should just be `Q` for both + and - exponents though? Then solver should sort out the difference between the two in the various `_grad_next/_prev` methods?

On the other hand, in many calculations one might need to take `exp.Q.dag() if exp.type == exp.types["+"] else exp.Q` which would be annoying if one is working with the exponents outside of the solver.

Reply:
We ended up going with making `exp.Q` always the bath coupling operator and having the HEOMSolver calculate the daggered operators for `+` modes where needed.